cranelift/codegen/src/isa/x64/inst/mod.rs

```diff
@@ -2362,7 +2363,17 @@ fn x64_get_operands<F: Fn(VReg) -> VReg>(inst: &Inst, collector: &mut OperandCol
 
         Inst::ReturnCallUnknown { callee, info } => {
             let ReturnCallInfo { uses, tmp, .. } = &**info;
-            callee.get_operands(collector);
+            match callee {
+                RegMem::Reg { reg } => {
+                    // TODO(https://github.com/bytecodealliance/regalloc2/issues/145):
+                    // This shouldn't be a fixed register constraint, but it's not clear how to
+                    // pick a register that won't be clobbered by the callee-save restore code
+                    // emitted with a return_call_indirect. r10 is caller-saved, so this should be
+                    // safe to use.
+                    collector.reg_fixed_use(*reg, regs::r10())
+                }
+                RegMem::Mem { addr } => addr.get_operands(collector),
+            }
             collector.reg_early_def(tmp.to_writable_reg());
```

Comment:
I think that we could switch back to always using `r11` for the temp, as this change means we'll no longer accidentally pick `r11` for the destination.

Reply:
I don't think the `RegMem::Mem` case is safe here, because any registers in the address mode could themselves be callee-saved. We might need to require that `ReturnCallUnknown` is used only with a register, never a memory operand, unlike `CallUnknown`.

As for using `r11` for the temporary register, I think we actually must do that. In general I'd rather let RA2 choose the temporary, but it's used after callee-saved registers are restored, so as it stands now we could clobber something we're supposed to have preserved.

Reply:
I was wondering about this, especially since it's not clear to me why this is a `RegMem` and not a plain old `Reg`.

Reply:
Actually, switching it to accepting a `Reg` was really easy, and the only place we construct this instruction we were always providing a register that we turned into a `RegMem` :+1:

Reply:
In general it makes sense to me to allow an indirect call on x86 to sink a load, if that's where the callee came from.

We could even hit this case from Wasmtime, where an indirect call through a table gets the callee function pointer by loading it out of a struct in memory. (I think load sinking currently fails there because there's another load in between; we might be able to micro-optimize this for x86 by swapping the two loads.)

However, I think that's a code-size optimization we just can't safely support for tail calls.